File: workflow/scripts/patch_raster.R
library("rgrass")

gisbase <- system2("grass", args = c("--config", "path"), stdout = TRUE)
template <- terra::rast(matrix(), crs = "epsg:3067")
invisible(initGRASS(gisBase = gisbase, home = tempdir(), SG = template))

input_rasters <- snakemake@input

output_field <- snakemake@wildcards[["gridcell_field"]]

field_type <- snakemake@params[["field_type"]]
resolution <- snakemake@params[["resolution"]]


if (field_type == "category") {
    gdal_type <- "Byte"
    gdal_nodata <- 0
    geotiff_predictor <- 2
    nbits <- NULL
} else if (field_type %in% c("integer", "date")) {
    if (substr(output_field, 1, 9) == "stemcount") {
        gdal_type <- "UInt32"
        gdal_nodata <- 2^20 - 1
        nbits <- "NBITS=20"
        geotiff_predictor <- 1
    } else {
        gdal_type <- "UInt16"
        gdal_nodata <- 2^16 - 1
        nbits <- NULL
        geotiff_predictor <- 2
    }
} else if (field_type == "real") {
    gdal_type <- "Float32"
    gdal_nodata <- -65504
    geotiff_predictor <- 3
    nbits <- "NBITS=16"
}

r_seq <- seq_along(input_rasters)

for (r in r_seq) {
    r_name <- paste0("r.", r)

    execGRASS(
        "r.in.gdal",
        flags = c("o"),
        input = input_rasters[[r]],
        output = r_name
    )
}

r_seq <- paste0("r.", r_seq)

execGRASS(
    "g.region",
    raster = r_seq
)

execGRASS(
    "r.patch",
    input=r_seq,
    output=output_field
)

if (output_field == "developmentclass") {
    category_file <- tempfile()

    writeLines(
        c("1:A0", "2:S0", "3:T1", "4:T2", "5:Y1", "6:ER", "7:02", "8:03", "9:04", "10:05"),
        con = category_file
    )

    execGRASS(
        "r.category",
        map=output_field,
        separator=":",
        rules=category_file
    )

    flags <- c("t")
} else {
    flags <- NULL
}

execGRASS(
    "r.out.gdal",
    flags = c("c", "f", flags),
    input = output_field,
    output = snakemake@output[[1]],
    type = gdal_type,
    nodata = gdal_nodata,
    createopt = c("COMPRESS=LZW", paste0("PREDICTOR=", geotiff_predictor), nbits, "TILED=YES", "BIGTIFF=YES")
)

File: workflow/scripts/gridcell/patch_raster.py
from snakemake.script import snakemake

import subprocess
import sys

from os.path import join
from tempfile import TemporaryDirectory, NamedTemporaryFile

sys.path.append(
    subprocess.check_output(["grass", "--config", "python_path"], text=True).strip()
)

from grass.script.core import create_project
from grass.script.setup import init as init_grass
from grass.experimental import TemporaryMapsetSession
from grass.tools import Tools

with TemporaryDirectory() as gisdb:
    create_project(path=join(gisdb, "project"), epsg=3067)

    with init_grass(gisdb, "project", "PERMANENT") as session:

        field_type = snakemake.params.field_type
        output_field = snakemake.wildcards.gridcell_field

        export_config = dict()

        match field_type:

            case "category" | "integer" | "date":
                grass_type = "CELL"

                match field_type:
                    case "category":
                        export_config |= {
                            "type": "Byte",
                            "nodata": 0,
                            "createopt": ["PREDICTOR=2"],
                        }

                    case "integer" if output_field[:9] == "stemcount":
                        export_config |= {
                            "type": "UInt32",
                            "nodata": 2**20 - 1,
                            "createopt": ["NBITS=20"],
                        }

                    case _:
                        export_config |= {
                            "type": "UInt16",
                            "nodata": 2**16 - 1,
                            "createopt": ["PREDICTOR=2"],
                        }

            case "real":
                grass_type = "FCELL"

                export_config |= {
                    "type": "Float32",
                    "nodata": -65504,
                    "createopt": [
                        "PREDICTOR=3",
                        "NBITS=16",
                    ],
                }

        export_config["createopt"] += ["COMPRESS=DEFLATE", "TILED=YES", "BIGTIFF=YES"]

        tools = Tools(quiet=True)

        input_rasters = {f"r.{i}": v for i, v in enumerate(snakemake.input)}

        for k, v in input_rasters.items():
            tools.r_external(
                flags="o",
                input=v,
                output=k,
            )

        tools.g_region(raster=input_rasters)

        tools.r_buildvrt(input=input_rasters, output=output_field)

        if output_field == "developmentclass":
            with NamedTemporaryFile(
                mode="w+",
                encoding="ascii",
                delete=False,
                dir=gisdb,
            ) as category_file:
                category_file.writelines(
                    f"{x}\n"
                    for x in [
                        "1:A0",
                        "2:S0",
                        "3:T1",
                        "4:T2",
                        "5:Y1",
                        "6:ER",
                        "7:02",
                        "8:03",
                        "9:04",
                        "10:05",
                    ]
                )
                category_file.close()

                tools.r_category(
                    map=output_field, separator=":", rules=category_file.name
                )

            flags = "cft"
        else:
            flags = "cf"

        tools.r_out_gdal(
            flags=flags, input=output_field, output=snakemake.output[0], **export_config
        )
